### PR TITLE
Fix issues found by PVS-Studio static analyzer #2

### PR DESCRIPTION
**Pull request by dimhotepus**
https://pvs-studio.com/en/order/open-source-license/

### CODE REVIEW

File: Rtp/RFC3550.cs [hunk 1]
```diff
@@ -180,8 +180,8 @@ public static IEnumerable<byte> ToCompoundBytes(IEnumerable<RtcpPacket> packets)
                     hasSourceDescription = true;
 
                     //if not already checked for a cname check now
-                    if (false == hasCName && packet.BlockCount > 0) using (SourceDescriptionReport asReport = new SourceDescriptionReport(packet, false)) if ((hasCName = asReport.HasCName)) break;
-                }                
+                    if (packet.BlockCount > 0) using (SourceDescriptionReport asReport = new SourceDescriptionReport(packet, false)) if ((hasCName = asReport.HasCName)) break;
```
Comment (juliusfriedman) [inline Rtp/RFC3550.cs:183]: Why drop the check for cName here?

File: Rtp/RFC3550.cs [hunk 2]
```diff
@@ -348,7 +348,7 @@ public static IEnumerable<RtcpPacket> FromCompoundBytes(byte[] array, int offset
                     hasSourceDescription = true;
 
                     //if not already checked for a cname check now
-                    if (false == hasCName && currentPacket.BlockCount > 0) using (SourceDescriptionReport asReport = new SourceDescriptionReport(currentPacket, false)) if ((hasCName = asReport.HasCName)) break;
+                    if (currentPacket.BlockCount > 0) using (SourceDescriptionReport asReport = new SourceDescriptionReport(currentPacket, false)) if ((hasCName = asReport.HasCName)) break;
```
Comment (juliusfriedman) [inline Rtp/RFC3550.cs:351]: Why would we drop the check for `hasCName`?

File: Rtp/RFC3550.cs [hunk 3]
```diff
@@ -438,7 +438,7 @@ public static int ReadPadding(byte[] buffer, int offset, int count)
         public static IEnumerable<byte> CreatePadding(int amount) //int? codeAmount
         {
             if (amount <= 0) return Enumerable.Empty<byte>();
-            if (amount > byte.MaxValue) Common.Binary.CreateOverflowException("amount", amount, byte.MinValue.ToString(), byte.MaxValue.ToString());
+            if (amount > byte.MaxValue) throw Common.Binary.CreateOverflowException("amount", amount, byte.MinValue.ToString(), byte.MaxValue.ToString());
```
Comment (juliusfriedman) [inline Rtp/RFC3550.cs:441]: This is a good fix by itself. It's a missing throw.

File: Rtp/RFC3550.cs [hunk 4]
```diff
@@ -1912,7 +1912,7 @@ internal class RFC3550UnitTests
     {
         public static void TestSourceList()
         {
-            Media.RFC3550.SourceList sourceList = new RFC3550.SourceList(0);
+            using Media.RFC3550.SourceList sourceList = new RFC3550.SourceList(0);
```
Comment (juliusfriedman) [inline Rtp/RFC3550.cs:1915]: Good fix

File: Rtp/Rtcp/RtcpHeader.cs [hunk 5]
```diff
@@ -225,7 +225,7 @@ public int LengthInWordsMinusOne
                 /*CheckDisposed();*/
 
                 //Write the value
-                if (value > RtcpHeader.MinimumLengthInWords) Binary.CreateOverflowException("LengthInWordsMinusOne", value, ushort.MinValue.ToString(), ushort.MaxValue.ToString());
+                if (value > RtcpHeader.MinimumLengthInWords) throw Binary.CreateOverflowException("LengthInWordsMinusOne", value, ushort.MinValue.ToString(), ushort.MaxValue.ToString());
```
Comment (juliusfriedman) [inline Rtp/Rtcp/RtcpHeader.cs:228]: Good fix

File: Rtp/Rtcp/RtcpPacket.cs [hunk 6]
```diff
@@ -307,7 +307,7 @@ public RtcpPacket(int version, int payloadType, int padding, int ssrc, int block
             :base(shouldDispose)
         {
             //If the padding is greater than allow throw an overflow exception
-            if (padding < 0 || padding > byte.MaxValue) Binary.CreateOverflowException("padding", padding, byte.MinValue.ToString(), byte.MaxValue.ToString());            
+            if (padding < 0 || padding > byte.MaxValue) throw Binary.CreateOverflowException("padding", padding, byte.MinValue.ToString(), byte.MaxValue.ToString());
```
Comment (juliusfriedman) [inline Rtp/Rtcp/RtcpPacket.cs:310]: Good fix.

File: Rtp/Rtcp/RtcpReport.cs [hunk 7]
```diff
@@ -323,14 +323,12 @@ public virtual void Add(IReportBlock reportBlock)
 
         public virtual bool Remove(IReportBlock reportBlock)
         {
-            if (reportBlock == null) return false;
+            if (reportBlock == null) throw new ArgumentNullException("reportBlock");
```
Comment (juliusfriedman) [inline Rtp/Rtcp/RtcpReport.cs:326]: I don't think I agree with this fix.

File: Rtp/Rtcp/SourceDescriptionReport.cs [hunk 8]
```diff
@@ -316,7 +316,7 @@ public SourceDescriptionItem(SourceDescriptionItemType itemType, IEnumerable<byt
                 int octetCount = octets.Count();
 
                 //If the amount of contained octets is greater than the maximum value allowed then throw an overflow exception.
-                if (octetCount > byte.MaxValue) Binary.CreateOverflowException("octets", octetCount, byte.MinValue.ToString(), byte.MaxValue.ToString());
+                if (octetCount > byte.MaxValue) throw Binary.CreateOverflowException("octets", octetCount, byte.MinValue.ToString(), byte.MaxValue.ToString());
```
Comment (juliusfriedman) [inline Rtp/Rtcp/SourceDescriptionReport.cs:320]: Good Fix!

File: Rtp/Rtcp/SourceDescriptionReport.cs [hunk 9]
```diff
@@ -1253,7 +1253,7 @@ public bool Remove(SourceDescriptionChunk chunk)
             foreach (SourceDescriptionChunk currentChunk in GetChunkIterator())
             {
                 //If the chunk yielded in the iterator matches the chunk identifier a match is present.
-                if (contained = chunk.ChunkData == currentChunk.ChunkData)
+                if (contained = (chunk.ChunkData == currentChunk.ChunkData))
                 {
```
Comment (juliusfriedman) [inline Rtp/Rtcp/SourceDescriptionReport.cs:1257]: Are the parenthesis that important?

File: Rtp/RtpClient.Methods.cs [hunk 10]
```diff
@@ -358,7 +358,7 @@ public System.Collections.Generic.IEnumerable<Rtcp.RtcpReport> PrepareReports(Tr
         internal protected /*virtual*/ int SendReceiversReport(TransportContext context, bool force = false)
         {
             //Determine if the ReceiversReport can be sent.
-            if (Common.IDisposedExtensions.IsNullOrDisposed(this) | Common.IDisposedExtensions.IsNullOrDisposed(context)  //If the context is disposed
+            if (Common.IDisposedExtensions.IsNullOrDisposed(this) || Common.IDisposedExtensions.IsNullOrDisposed(context)  //If the context is disposed
```
Comment (juliusfriedman) [inline Rtp/RtpClient.Methods.cs:361]: I suppose this is a good fix, the other way was intentionally checking in parallel. but this is good I suppose.

File: Rtp/RtpClient.Methods.cs [hunk 11]
```diff
@@ -493,7 +493,7 @@ public System.Collections.Generic.IEnumerable<Rtcp.RtcpReport> PrepareReports(Tr
             if (Common.IDisposedExtensions.IsNullOrDisposed(this) || packets == null) return 0;
 
             //If we don't have an transportContext to send on or the transportContext has not been identified or Rtcp is Disabled or there is no remote rtcp end point
-            if (Common.IDisposedExtensions.IsNullOrDisposed(context) || context.SynchronizationSourceIdentifier.Equals(Common.Binary.Zero) | false.Equals(context.IsRtcpEnabled) | context.RemoteRtcp == null)
+            if (Common.IDisposedExtensions.IsNullOrDisposed(context) || context.SynchronizationSourceIdentifier.Equals(Common.Binary.Zero) || false.Equals(context.IsRtcpEnabled) || context.RemoteRtcp == null)
```
Comment (juliusfriedman) [inline Rtp/RtpClient.Methods.cs]: These are all intentional.

**Review by juliusfriedman — CHANGES_REQUESTED** [review]
I put my comments on what is intentional vs what is not intentional, please let me know how strongly you feel and I will review again from there.

File: Rtp/RFC3550.cs (hunk 1 above)
Comment (juliusfriedman) [reply Rtp/RFC3550.cs:183]: I see its a duplicate check.

File: Rtp/RFC3550.cs (hunk 2 above)
Comment (juliusfriedman) [reply Rtp/RFC3550.cs:351]: I see this is a duplicate check...

File: Rtp/Rtcp/RtcpReport.cs (hunk 7 above)
Comment (juliusfriedman) [reply Rtp/Rtcp/RtcpReport.cs:326]: After review I see its checked twice...

**Comment by dimhotepus** [comment]
@juliusfriedman Could you please recheck?

**Review by juliusfriedman — APPROVED** [review]

**Comment by juliusfriedman** [comment]
Thank you for your contributions!!!

**Pull request merged**